### ISSUE TITLE
Redirect option in boot files does not work

### ISSUE THREAD

**Issue by ianlchapman-noodlelive** [quasarframework/quasar#10592]
**Describe the bug**
When using the redirect option in boot files the redirect does not work.

**To Reproduce**
1. Create a new quasar project via the CLI
2. Create a new boot file via the CLI
3. Copy the code below in to the boot file
4. Add the boot file to quasar.conf.js
5. Add an extra route to routes.js so that '/#/1' leads somewhere
6. Visit any url route except '/#/1'
7. Verify the url changes - however the component does not load - just a blank white screen
8. Also note that the redirect does not handle hash mode router either

Boot file
```javascript
import { boot } from 'quasar/wrappers'

// "async" is optional;
// more info on params: https://v2.quasar.dev/quasar-cli/boot-files
export default boot(async ( { redirect, urlPath } ) => {
  if(urlPath !== '/#/1') {
    console.log('before redirect');
    redirect('/#/1');
    console.log('after redirect');
  }
})
```

Routes.js
```javascript
  {
    path: '/1',
    component: () => import('layouts/MainLayout.vue'),
    children: [
      { path: '', component: () => import('pages/Index.vue') }
    ]
  },
  {
    path: '/',
    component: () => import('layouts/MainLayout.vue'),
    children: [
      { path: '', component: () => import('pages/Index.vue') }
    ]
  },
```

**Expected behavior**
1. Redirect should redirect to the new route and load the component
2. Redirect should handle hash mode router correctly

**Screenshots**
N/A

**Platform (please complete the following information):**
Quasar Version:
@quasar/app Version: 3.1.0

Quasar mode:
  - [*] SPA
  - [?] SSR
  - [?] PWA
  - [?] Electron
  - [*] Cordova
  - [?] Capacitor
  - [?] BEX

Tested on:
  - [*] SPA
  - [?] SSR
  - [?] PWA
  - [?] Electron
  - [*] Cordova
  - [?] Capacitor
  - [?] BEX

OS:
Node: 16.8.0
NPM: 7.21.0
Yarn: N/A
Browsers: Chrome & Safari
iOS: untested
Android: untested
Electron: untested

**Comment by ianlchapman-noodlelive** [comment]
Also probably a related issue:

https://github.com/quasarframework/quasar/issues/10062

**Comment by ianlchapman-noodlelive** [comment]
To further add to this ticket following up from Discord:

Changing the path in redirect has the following outcomes:

`redirect('/1');` - Does not honour the hash mode in the router - therefore redirects to '/1' which is not a valid route
`redirect({ path: '/1' });` - Does not honour the hash mode in the router - therefore redirects to '/1' which is not a valid route
`redirect('/#/1');` - Honours the hash mode in the router - redirects to the correct page - blank white screen - no errors in console
`redirect({ path: '/#/1' });` - Gets stuck in an infinite redirection loop

**Comment by ianlchapman-noodlelive** [comment]
So after some investigations I have found the issue. The issue is caused by `.quasar/client-entry.js` which declares the following function:

```javascript
let hasRedirected = false
  const redirect = url => {
    hasRedirected = true
    const normalized = Object(url) === url
      ? router.resolve(url).fullPath
      : url

    window.location.href = normalized
  }
```

The table below outlines the pass/fail functionality of this code block for the usage capabilities outlined in the documentation:

|Router Mode|Full url (http...)|Relative url (/etc)|Router Object (path)|Router Object(named)|
|---|---|---|---|---|
|History|PASS|PASS|PASS|PASS|
|Hash (Default created by Quasar CLI)|PASS|FAIL|FAIL|FAIL|

Below is some code I have written which in my test app fixes the issue and makes it work.

```javascript
  let hasRedirected = false
  const redirect = url => {
    hasRedirected = true

    if(typeof url === 'string' && url[0] !== '/') {
      window.location.href = url;
      return;
    }

    window.location.href = router.resolve(url).href;
    window.location.reload();
  }
```
Note that on the `router.resolve` calls we need to use the href return not the fullPath. Also in cases where a string url is returned we have to check if it is absolute or relative as relative paths will need to be passed through `router.resolve` to get the correct hash attached. Alternatively a full '/#/etc' path can be passed but this should be documented so everyone understands.

Note I have needed to add the `window.location.reload()` function call as in my hash mode scenario the page doesn't seem to load correctly if I don't. This might be avoided by doing a programatic `router.push` but I don't know enough about the internals to know if this approach would work. Trying `router.push` in the boot file seems to complete the redirect but never mounts the app. This is possibly caused by the:

```javascript
  if (hasRedirected === true) {
    return
  }
```

In `client-entry.js` which prevents the app from mounting.


Furthermore further down in the promise version of the functionality the following code is used:

```javascript
try {
      await bootFiles[i]({
        app,
        router,
        
        ssrContext: null,
        redirect,
        urlPath,
        publicPath
      })
    }
    catch (err) {
      if (err && err.url) {
        window.location.href = err.url
        return
      }

      console.error('[Quasar] boot error:', err)
      return
    }
```

This code doesn't respect any of the router specifications at all. Therefore I would propose that the same redirect function is used here.

```javascript
      if (err && err.url) {
        redirect(err.url);
        return
      }
```

This client-entry.js file is generated by quasar cli automatically when running the command and is overwritten each time. I am not familiar enough with have any of that side of the tooling works to create a fix.

I believe that this same fix should also work for this issue https://github.com/quasarframework/quasar/issues/10062 which seems to be stuck in a similar situation.

**Comment by rstoenescu** [comment]
Fix for q/app will be available in "@quasar/app" v3.1.4

The following were invalid redirects to begin with: `redirect('/#/1')`, `redirect({ path: '/#/1' })`. But I agree that it was not clear from the docs. So improved them.

Extract from the improved docs (which had this section but it wasn't so detailed):

---

### Redirecting to another page

::: warning
Please be mindful when redirecting as you might configure the app to go into an infinite redirect loop.
:::

```js
export default ({ urlPath, redirect }) => {
  // ...
  const isAuthorized = // ...
  if (!isAuthorized && !urlPath.startsWith('/login')) {
    redirect({ path: '/login' })
    return
  }
  // ...
}
```

The `redirect()` method accepts a String (full URL) or a Vue Router location String or Object. On SSR it can receive a second parameter which should be a Number for any of the HTTP STATUS codes that redirect the browser (3xx ones).

```js
// Examples for redirect() with a Vue Router location:
redirect('/1') // Vue Router location as String
redirect({ path: '/1' }) // Vue Router location as Object

// Example for redirect() with a URL:
redirect('https://quasar.dev')
```

::: warning IMPORTANT!
The Vue Router location (in String or Object form) does not refer to URL path (and hash), but to the actual Vue Router routes that you have defined.
So **don't add the publicPath** to it and if you're using the Vue Router hash mode then don't add the hash to it.

<br>Let's say that we have this Vue Router route defined:<br><br>

```js
{
  path: '/one',
  component: PageOne
}
```

<br>Then **regardless of our publicPath** we can call `redirect()` like this:<br><br>

```js
// publicPath: /wiki; vueRouterMode: history
redirect('/one') // good way
redirect({ path: '/one' }) // good way
redirect('/wiki/one') // WRONG!

// publicPath: /wiki; vueRouterMode: hash
redirect('/one') // good way
redirect({ path: '/one' }) // good way
redirect('/wiki/#/one') // WRONG!

// no publicPath; vueRouterMode: hash
redirect('/one') // good way
redirect({ path: '/one' }) // good way
redirect('/#/one') // WRONG!
```
:::

As it was mentioned in the previous sections, the default export of a boot file can return a Promise. If this Promise gets rejected with an Object that contains a "url" property, then Quasar CLI will redirect the user to that URL:

```js
export default ({ urlPath }) => {
  return new Promise((resolve, reject) => {
    // ...
    const isAuthorized = // ...
    if (!isAuthorized && !urlPath.startsWith('/login')) {
      // the "url" param here is of the same type
      // as for "redirect" above
      reject({ url: '/login' })
      return
    }
    // ...
  })
}
```

Or a simpler equivalent:

```js
export default () => {
  // ...
  const isAuthorized = // ...
  if (!isAuthorized && !urlPath.startsWith('/login')) {
    return Promise.reject({ url: '/login' })
  }
  // ...
}
```
